lib/linux_admin/ssh.rb

```diff
@@ -0,0 +1,96 @@
+module LinuxAdmin
+  class SSH
+    attr_reader :ip
+    attr_reader :username
+    attr_reader :private_key
+    attr_reader :agent
+
+    def initialize(ip, username, private_key)
+      @ip = ip
+      @private_key = private_key
+      @username = username
+    end
+
+    def perform_commands(commands = [], options = {})
+      result = nil
+      if options[:agent_enabled]
+        LinuxAdmin::SSHAgent.with_service(@private_key, options[:agent_socket]) do |socket|
+          result = if block_given?
+                     execute_commands(commands, :socket => socket, :stdin => options[:stdin], &Proc.new)
+                   else
+                     execute_commands(commands, :socket => socket, :stdin => options[:stdin])
+                   end
+        end
+      else
+        result = if block_given?
+                   execute_commands(commands, :stdin => options[:stdin], &Proc.new)
+                 else
+                   execute_commands(commands, :stdin => options[:stdin])
+                 end
+      end
+      result
+    end
+
+    private
+
+    def execute_commands(commands = [], options = {})
+      result = nil
+      args = {}
+      if options[:socket]
+        args.merge!(:paranoid                   => false,
+                    :forward_agent              => true,
+                    :number_of_password_prompts => 0,
+                    :agent_socket_factory       => -> { UNIXSocket.open(options[:socket]) })
+      else
+        args[:keys] = [@private_key]
+      end
+      Net::SSH.start(@ip, @username, args) do |ssh|
```

Comment:
I think I'd prefer to see the `require net/ssh` right before the call rather than at the top of the file.  Otherwise everyone that requires `linux_admin` also gets `net/ssh` loaded even if they don't plan to use it.

In a large app like ManageIQ/manageiq we try to delay load as much as possible to reduce the memory footprint.  i.e. There's no reason for some worker to load `net/ssh`.